velox/connectors/hive/TableFormatReader.h

```diff
@@ -0,0 +1,127 @@
+#include "velox/dwio/common/BufferedInput.h"
+#include "velox/dwio/common/Reader.h"
+
+namespace facebook::velox::connector::hive {
+
+static const char* kPath = "$path";
+static const char* kBucket = "$bucket";
+
+class TableFormatReader {
+ public:
+  static std::unique_ptr<TableFormatReader> create(
+      std::shared_ptr<velox::connector::hive::HiveConnectorSplit> hiveSplit,
+      RowTypePtr readerOutputType,
+      std::unordered_map<std::string, std::shared_ptr<HiveColumnHandle>>&
+          partitionKeys,
+      std::shared_ptr<common::ScanSpec> scanSpec,
+      memory::MemoryPool*& pool);
+
+  TableFormatReader(
+      std::shared_ptr<velox::connector::hive::HiveConnectorSplit> hiveSplit,
+      RowTypePtr readerOutputType,
+      std::unordered_map<std::string, std::shared_ptr<HiveColumnHandle>>&
+          partitionKeys,
+      std::shared_ptr<common::ScanSpec> scanSpec,
+      memory::MemoryPool*& pool);
+
+  virtual ~TableFormatReader() = default;
+
+  /**
+   * This function is used by different table formats like Iceberg and Hudi to
+   * do additional preparations before reading the split, e.g. Open delete files
+   * or log files, and add column adapatations for metadata columns
+   */
+  virtual void prepareSplit(
+      const std::shared_ptr<HiveTableHandle>& hiveTableHandle,
+      const dwio::common::ReaderOptions& readerOptions,
+      std::unique_ptr<dwio::common::BufferedInput> baseFileInput,
+      std::shared_ptr<common::MetadataFilter> metadataFilter,
+      dwio::common::RuntimeStatistics& runtimeStats);
+
+  virtual uint64_t next(int64_t size, VectorPtr& output);
+
+  std::unique_ptr<dwio::common::RowReader> createRowReader(
+      dwio::common::RowReaderOptions& options) {
+    return baseReader_->createRowReader(options);
+  };
+
+  std::shared_ptr<dwio::common::RowReader> baseRowReader() {
+    return baseRowReader_;
+  }
+
+  void resetFilterCaches();
+
+  bool emptySplit() const;
+
+  void resetSplit();
+
+  int64_t estimatedRowSize() const;
+
+  void updateRuntimeStats(dwio::common::RuntimeStatistics& stats) const;
+
+  bool allPrefetchIssued() const;
+
+  std::string toString() const;
+
+ protected:
+  /**
+   * Different table formats may have different meatadata columns. This function
+   * will be used to update the scanSpec for these columns.
+   */
+  virtual std::vector<TypePtr> adaptColumns(
+      const RowTypePtr& fileType,
+      const std::shared_ptr<const velox::RowType>& tableSchema);
+
+  void setConstantValue(
+      common::ScanSpec* FOLLY_NONNULL spec,
+      const TypePtr& type,
+      const velox::variant& value) const;
+
+  void setNullConstantValue(
+      common::ScanSpec* FOLLY_NONNULL spec,
+      const TypePtr& type) const;
+
+  void setPartitionValue(
+      common::ScanSpec* FOLLY_NONNULL spec,
+      const std::string& partitionKey,
+      const std::optional<std::string>& value) const;
+
+  std::shared_ptr<HiveConnectorSplit> hiveSplit_;
```

Comment:
Can we keep these fields private?

Reply:
The subclasses will need to further parse the format specific information in the split

Reply:
I moved rowReaderOpts_ to be private, but other fields will be used by IcebergTableReader.